contracts/RMRK/extension/secureTokenTransferProtocol/RMRKSecureTokenTransferProtocol.sol

```diff
@@ -0,0 +1,312 @@
+// SPDX-License-Identifier: Apache-2.0
+
+pragma solidity ^0.8.18;
+
+import "./IRMRKSecureTokenTransferProtocol.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+import "@openzeppelin/contracts/token/ERC1155/IERC1155.sol";
+
+error InvalidValue();
+error InvalidAddress();
+error InsufficientBalance();
+
+abstract contract RMRKSecureTokenTransferProtocol is
+    IRMRKSecureTokenTransferProtocol
+{
+    mapping(uint256 tokenId => mapping(address tokenAddress => mapping(TokenType tokenType => mapping(uint256 heldTokenId => uint256 balance))))
+        private _balances;
+
+    /**
+     * @inheritdoc IRMRKSecureTokenTransferProtocol
+     */
+    function balanceOfToken(
+        address tokenContract,
+        TokenType tokenType,
+        uint256 tokenId,
+        uint256 heldTokenId
+    ) external view returns (uint256) {
+        if (tokenType == TokenType.ERC20) {
+            return _balances[tokenId][tokenContract][tokenType][heldTokenId];
+        } else {
+            return _balances[tokenId][tokenContract][tokenType][heldTokenId];
+        }
+    }
+
+    /**
+     * @notice Transfer held tokens from a specific token.
+     * @dev The balance MUST be transferred from this smart contract.
+     * @dev Implementers should validate that the `msg.sender` is either the token owner or approved to manage it before
+     *  calling this.
+     * @dev If the token type is `ERC-20`, the `heldTokenId` MUST be ignored.
+     * @dev IF the token type is `ERC-721`, the `amount` MUST be ignored.
+     * @param tokenContract The address of the held token's smart contract
+     * @param tokenType The type of the token being transferred
+     * @param tokenId The ID of the token to transfer the held token from
+     * @param heldTokenId The ID of the held token to transfer
+     * @param amount The number of held tokens to transfer
+     * @param to The address to transfer the held tokens to
+     * @param data Additional data with no specified format, to allow for custom logic
+     */
+    function _transferHeldTokenFromToken(
+        address tokenContract,
+        TokenType tokenType,
+        uint256 tokenId,
+        uint256 heldTokenId,
+        uint256 amount,
+        address to,
+        bytes memory data
+    ) internal {
+        if (amount == 0 && tokenType != TokenType.ERC721) {
```

Comment:
Wouldn't it make more sense for amount to be always one for 721s?

Reply:
Actually it has to be 1, you are increasing the balance by 1 when receiving but decreasing by zero when sending.

Reply:
Yeah, you are correct. This didn't conform to the parameters that needed to be ignored based on the token type. I fixed it in the most recent commit.